### PR TITLE
Support fetching without the --progress option

### LINKED ISSUE #894: Run git fetch with --quiet by default
https://git-scm.com/docs/git-fetch#Documentation/git-fetch.txt--q

As has been noted before in the prematurely closed #489 and #18 the fetch progress does not add any value and just clutters the logs, esp. the raw logs.
The folding in the UI does not fix the issue for the raw logs.
And in general it's a waste of bandwidth and everything.

### PR DESCRIPTION
**Pull request by simonbaird**
If this is merged, setting the `progress` option to false in the `with` section of the workflow step will cause git fetch to run without the `--progress` flag.

Note that the default value of this new option is `true`, which matches the current behavior, so this PR introduces no behavior change unless the new option is explicitly present and set to false.

The motivation is to be able to suppress the noisy progress status output which adds many hundreds of `remote: Counting objects: 85% (386/453)` and similar lines in the workflow log.

This should be sufficient to resolve #894 and its older friends, though the solution is different to the one proposed there because it doesn't use the `--quiet` flag. IIUC git doesn't show the progress status by default when running on GitHub since the output is not a terminal. Adding the `--progress` flag is what forces the progress output to be shown, so that's why removing it is all that's needed.

Adding the `--quiet` flag doesn't make a lot of difference once the `--progress` flag is removed, and actually I think using `--quiet` would suppress some other more useful output that would be better left visible.

### CODE REVIEW

**Comment by simonbaird** [comment]
I'm using it [here](https://github.com/simonbaird/tiddlyhost/actions/runs/3844090549/jobs/6546904868#step:2:50) and it seems to work well.

**Comment by simonbaird** [comment]
[This workflow run](https://github.com/simonbaird/tiddlyhost/actions/runs/3844120635/jobs/6546957340#step:2:50) shows it working with the `--progess output` which is the default.

**Comment by simonbaird** [comment]
@TingluoHuang would this be something you'd consider taking a look at? I'm looking at your comments from 2019 on #18.

**Comment by yeikel** [comment]
What is blocking this change? 

I am looking forward to seeing this released as well

**Comment by simonbaird** [comment]
Rebased on upstream/main, (currently f095bcc).

**Comment by simonbaird** [comment]
Passes formatting and lint checks for me locally:

```
for t in build format-check lint; do npm run $t; done

> checkout@3.5.2 build
> tsc && ncc build && node lib/misc/generate-docs.js

ncc: Version 0.36.1
ncc: Compiling file index.js into CJS
877kB  dist/index.js
877kB  [1343ms] - ncc 0.36.1

> checkout@3.5.2 format-check
> prettier --check '**/*.ts'

Checking formatting...
All matched files use Prettier code style!

> checkout@3.5.2 lint
> eslint src/**/*.ts
```

**Comment by simonbaird** [comment]
I reworded the PR description and added some addition explanations.

**Comment by simonbaird** [comment]
Rebased on 96f5310.

**Comment by simonbaird** [comment]
```
$ for t in build format-check lint; do npm run $t; done

> checkout@3.5.3 build
> tsc && ncc build && node lib/misc/generate-docs.js

ncc: Version 0.36.1
ncc: Compiling file index.js into CJS
879kB  dist/index.js
879kB  [1112ms] - ncc 0.36.1

> checkout@3.5.3 format-check
> prettier --check '**/*.ts'

Checking formatting...
All matched files use Prettier code style!

> checkout@3.5.3 lint
> eslint src/**/*.ts
```

**Comment by slang25** [comment]
@luketomlinson are you able to comment, there are a lot of people wanting this.

**Comment by yeikel** [comment]
Well, currently there are conflicts. It would help if we can resolve these first. Otherwise it'll take more time before this is looked at again 

@simonbaird Could you please work on that?

**Comment by simonbaird** [comment]
Rebased on upstream/main (currently at 97a652b80035363df47baee5031ec8670b8878ac). Now includes some test coverage :sunglasses: .

File: __test__/git-command-manager.test.ts [hunk 1]
```diff
@@ -263,4 +259,120 @@ describe('Test fetchDepth and fetchTags options', () => {
...
+      doSparseCheckout
+    )
+    const refSpec = ['refspec1', 'refspec2']
+    const options = {
+      filter: 'filterValue',
+      fetchDepth: 42,
+      showProgress: true
+    }
+
+    await git.fetch(refSpec, options)
```
Comment (simonbaird) [inline __test__/git-command-manager.test.ts:319]: It might be nice to merge options onto the default options here so the assertions become more like real world examples. Would be a separate commit though.

File: __test__/git-command-manager.test.ts [hunk 2]
```diff
@@ -263,4 +259,120 @@ describe('Test fetchDepth and fetchTags options', () => {
...
+        '--no-recurse-submodules',
+        '--progress',
+        '--filter=filterValue',
+        'origin',
+        'refspec1',
+        'refspec2'
+      ],
+      expect.any(Object)
+    )
+  })
```
Comment (simonbaird) [inline __test__/git-command-manager.test.ts:377]: I'd be happy to remove some of the test coverage if it's considered too much. I was trying to follow the example set in #579 .

**Review by vanZeben — APPROVED** [review]
Thanks for putting this together @simonbaird, especially with the comprehensive test coverage! 🎉  I'll cut a release for this later today

**Comment by vanZeben** [comment]
I lied, checking internally, we're planning on cutting a new Major version release next week to include the [recent node20 bump](https://github.com/actions/checkout/pull/1436), that major release will include this too, so look forward to `v4.0.0` for this change!

**Comment by yeikel** [comment]
> I lied, checking internally, we're planning on cutting a new Major version release next week to include the [recent node20 bump](https://github.com/actions/checkout/pull/1436), that major release will include this too, so look forward to `v4.0.0` for this change!

Can this be backported to 3.x as well?

I am running this in an environment where the node20 runners are not available yet

**Comment by Gakk** [comment]
@simonbaird Setting `show-progress: false` gets rid of the initial `Counting objects`, `Compressing objects`, `Receiving objects` and `Resolving deltas` under the step `Enumerating objects` 👍

But the step `Checking out the ref` still runs with `--progress` adding a lot of log lines for `Updating files:`

image

I get this when using `runs-on: [windows-latest]` and:

```yml
    - name: Checkout
      uses: actions/checkout@3df4ab11eba7bda6032a0b82a6bb43b11571feac # v4.0.0
      # https://github.com/actions/checkout/releases
      with:
        show-progress: false
        fetch-depth: 0
```

And also with `runs-on: [windows-2019]` and:

```yml
    - name: Checkout
      uses: actions/checkout@3df4ab11eba7bda6032a0b82a6bb43b11571feac # v4.0.0
      with:
        show-progress: false
```

**Comment by simonbaird** [comment]
You're 100% right, the changes from the `fetch` method need to be also done for the `checkout` method. I'll take a look. But let's file a new issue.

**Comment by simonbaird** [comment]
> You're 100% right, the changes from the `fetch` method need to be also done for the `checkout` method. I'll take a look. But let's file a new issue.

See #1455.

**Review by AngelicaOvesdotter — COMMENTED** [review]
🤩
